File: draft_design_pattern/template_method.py
# template methodパターンは、特定のアルゴリズムの実行手順の骨格部分を定義します。手順は子クラスで実装します
from abc import ABC, abstractmethod


class Builder(ABC):
    # Template method
    def build(self):
        self.test()
        self.lint()
        self.assemble()
        self.deploy()

    @abstractmethod
    def test(self) -> None:
        pass

    @abstractmethod
    def lint(self) -> None:
        pass

    @abstractmethod
    def assemble(self) -> None:
        pass

    @abstractmethod
    def deploy(self) -> None:
        pass


class AndroidBuilder(Builder):
    def test(self):
        print('Androidのテストを実行')

    def lint(self):
        print('AndroidコードのLintを実行')

    def assemble(self):
        print('Androidビルドのアセンブリを実行')

    def deploy(self):
        print('Androidビルドをサーバーにデプロイ')


class IosBuilder(Builder):
    def test(self):
        print('iOSのテストを実行')

    def lint(self):
        print('iOSコードのLintを実行')

    def assemble(self):
        print('iOSビルドのアセンブリを実行')

    def deploy(self):
        print('iOSビルドをサーバーにデプロイ')


androidBuilder = AndroidBuilder()
androidBuilder.build()

iosBuilder = IosBuilder()
iosBuilder.build()

File: draft_design_pattern/template_method.ts
export {};

abstract class Builder {
  // Template method
  build(): void {
    this.test();
    this.lint();
    this.assemble();
    this.deploy();
  }

  abstract test(): void;
  abstract lint(): void;
  abstract assemble(): void;
  abstract deploy(): void;
}

class AndroidBuilder extends Builder {
  test() {
    console.log("Androidのテストを実行");
  }

  lint() {
    console.log("AndroidコードのLintを実行");
  }

  assemble() {
    console.log("Androidビルドのアセンブリを実行");
  }

  deploy() {
    console.log("Androidビルドをサーバーにデプロイ");
  }
}

class IosBuilder extends Builder {
  test() {
    console.log("iOSのテストを実行");
  }

  lint() {
    console.log("iOSコードのLintを実行");
  }

  assemble() {
    console.log("iOSビルドのアセンブリを実行");
  }

  deploy() {
    console.log("iOSビルドをサーバーにデプロイ");
  }
}

const androidBuilder = new AndroidBuilder();
androidBuilder.build();

const iosBuilder = new IosBuilder();
iosBuilder.build();
